api/client/client.go

```diff
@@ -77,29 +77,19 @@ func New(cfg Config) (*Client, error) {
 }
 
 func (c *Client) connect(noPingCheck bool) error {
-	dialer := grpc.WithContextDialer(func(ctx context.Context, addr string) (net.Conn, error) {
-		if c.isClosed() {
-			return nil, trace.ConnectionProblem(nil, "client is closed")
-		}
-		conn, err := c.c.Dialer.DialContext(ctx, "tcp", addr)
-		if err != nil {
-			return nil, trace.ConnectionProblem(err, "failed to dial")
-		}
-		return conn, nil
-	})
-
 	// Attempt to load credentials from each provider, and keep the first valid credentials.
 	var errs []error
 	var err error
 	for i, provider := range c.c.CredentialsProviders {
 		if c.c.Credentials, err = provider.Load(); err != nil {
-			errs = append(errs, trace.Errorf("CredentialsProvider[%v]: %v", i, err))
+			errs = append(errs, trace.Errorf("CredentialsProvider[%v]: failed to load Credentials: %v", i, err))
 			continue
 		}
 
 		c.conn, err = grpc.Dial(
 			constants.APIDomain,
-			dialer,
+			c.grpcAuthDialer(),
+			c.grpcProxyDialer(),
```

Comment:
IIUC, `grpc.WithContextDialer` shouldn't be specified multiple times.
If you specify it multiple times, only the last one is used.